pkg/controller/command/command_upgrade.go

```diff
@@ -1,26 +1,45 @@
 package command
 
 import (
+	"context"
+
 	apps "k8s.io/api/apps/v1"
+	batch "k8s.io/api/batch/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	contrail "github.com/Juniper/contrail-operator/pkg/apis/contrail/v1alpha1"
+	"github.com/Juniper/contrail-operator/pkg/job"
 )
 
-func performUpgradeStepIfNeeded(commandCR *contrail.Command, currentDeployment *apps.Deployment, oldDeploymentSpec *apps.DeploymentSpec) {
-	if imageIsUpgraded(currentDeployment, oldDeploymentSpec) {
+func (r *ReconcileCommand) performUpgradeStepIfNeeded(commandCR *contrail.Command, currentDeployment *apps.Deployment, oldDeploymentSpec *apps.DeploymentSpec, configMapName string) error {
+	if imageIsUpgraded(currentDeployment, oldDeploymentSpec) && commandCR.Status.UpgradeState != contrail.CommandUpgrading {
 		commandCR.Status.UpgradeState = contrail.CommandShuttingDownBeforeUpgrade
 	}
 	switch commandCR.Status.UpgradeState {
 	case contrail.CommandShuttingDownBeforeUpgrade:
+		currentDeployment.Spec.Template.Spec.Containers = oldDeploymentSpec.Template.Spec.Containers
+		currentDeployment.Spec.Template.Spec.InitContainers = oldDeploymentSpec.Template.Spec.InitContainers
+		currentDeployment.Spec.Replicas = int32ToPtr(0)
 		if currentDeployment.Status.Replicas == 0 {
-			currentDeployment.Spec.Replicas = commandCR.Spec.CommonConfiguration.Replicas
-			commandCR.Status.UpgradeState = contrail.CommandStartingUpgradedDeployment
-		} else {
-			currentDeployment.Spec.Template.Spec.Containers = oldDeploymentSpec.Template.Spec.Containers
-			currentDeployment.Spec.Template.Spec.InitContainers = oldDeploymentSpec.Template.Spec.InitContainers
-			currentDeployment.Spec.Replicas = int32ToPtr(0)
+			commandCR.Status.UpgradeState = contrail.CommandUpgrading
+		}
+	case contrail.CommandUpgrading:
+		currentDeployment.Spec.Replicas = int32ToPtr(0)
+		oldImage := getOldCommandImage(oldDeploymentSpec.Template.Spec.Containers, "command")
```

Comment:
Is it possible to change name of the container is deployment and container is CR?

Reply:
unified